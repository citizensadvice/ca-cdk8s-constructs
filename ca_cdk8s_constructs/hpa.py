from numbers import Number
import cdk8s_plus_31 as kplus
from constructs import Construct


def ca_hpa(
    scope: Construct,
    id: str,
    target: kplus.IScalable,
    cpu_utilization_target: Number = 75,
    memory_utilization_target: Number = 75,
    max_replicas: int = 3,
    min_replicas: int = 1,
) -> kplus.HorizontalPodAutoscaler:
    """Returns a HorizontalPodAutoscaler for a target.

    The utilization targets are applied to the aggregate CPU and memory usage of all containers in the target workload.

    Args:
        scope: The scope of the construct.
        target: The target to autoscale.
        cpu_utilization_target: The CPU utilization to target. Must be between 50 and 90. Defaults to 75.
        memory_utilization_target: The memory utilization to target. Must be between 50 and 90. Defaults to 75.
        max_replicas: The maximum number of replicas. Defaults to 3.
        min_replicas: The minimum number of replicas. Defaults to 1.

    Returns:
        A HorizontalPodAutoscaler.
    """
    if (
        cpu_utilization_target > 90
        or cpu_utilization_target < 50
        or memory_utilization_target > 90
        or memory_utilization_target < 50
    ):
        raise ValueError("Utilization values must be between 1 and 100")

    return kplus.HorizontalPodAutoscaler(
        scope,
        id,
        max_replicas=max_replicas,
        min_replicas=min_replicas,
        target=target,
        metrics=[
            kplus.Metric.resource_cpu(
                target=kplus.MetricTarget.average_utilization(cpu_utilization_target)
            ),
            kplus.Metric.resource_memory(
                target=kplus.MetricTarget.average_utilization(memory_utilization_target)
            ),
        ],
    )
