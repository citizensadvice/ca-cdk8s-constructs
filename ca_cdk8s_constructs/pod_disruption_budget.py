from cdk8s_plus_32.k8s import (
    KubePodDisruptionBudget,
    PodDisruptionBudgetSpec,
    LabelSelector,
    IntOrString,
)
from cdk8s_plus_32 import IScalable
from constructs import Construct


def ca_pod_disruption_budget(
    scope: Construct,
    id: str,
    target: IScalable,
    max_unavailable: int = 1,
) -> KubePodDisruptionBudget:
    """Create a PodDisruptionBudget for a deployment.

    The match_labels can be extracted from a Deployment or StatefulSet using the `match_labels` property.

    Args:
        scope: The scope of the construct.
        id: ID for the construct.
        target: The target to create a PodDisruptionBudget for. Usually a Deployment or StatefulSet.
        max_unavailable: The maximum number of unavailable pods. Defaults to 1.

    Returns:
        A PodDisruptionBudget.
    """
    return KubePodDisruptionBudget(
        scope,
        id,
        spec=PodDisruptionBudgetSpec(
            max_unavailable=IntOrString.from_number(max_unavailable),
            selector=LabelSelector(match_labels=target.match_labels),
        ),
    )
