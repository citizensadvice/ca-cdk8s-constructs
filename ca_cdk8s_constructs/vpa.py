from typing import Literal
import cdk8s_plus_32 as kplus
from constructs import Construct
from ca_cdk8s_constructs.imports.io.k8s.autoscaling import (
    VerticalPodAutoscaler,
    VerticalPodAutoscalerSpec,
    VerticalPodAutoscalerSpecTargetRef,
    VerticalPodAutoscalerSpecUpdatePolicy,
    VerticalPodAutoscalerSpecUpdatePolicyUpdateMode,
    VerticalPodAutoscalerSpecResourcePolicy,
    VerticalPodAutoscalerSpecResourcePolicyContainerPolicies,
    VerticalPodAutoscalerSpecResourcePolicyContainerPoliciesMinAllowed,
)


def ca_vpa(
    scope: Construct,
    id: str,
    target: kplus.IScalable,
    min_allowed_cpu: int = 100,
    update_mode: Literal["auto", "off", "initial", "recreate"] = "auto",
) -> VerticalPodAutoscaler:
    """Returns a VerticalPodAutoscaler for a target.

    The different modes are:
    - Off: VPA will only provide the recommendations, and it will not automatically change resource requirements.
    - Initial: VPA only assigns resource requests on pod creation and never changes them later.
    - Recreate: VPA assigns resource requests on pod creation time and updates them on existing pods by evicting and recreating them.
    - Auto: It recreates the pod based on the recommendation.

    The minimum allowed CPU is the minimum CPU that the VPA will assign to the pod. This is applied to all containers in the pod.

    Args:
        scope: The scope of the construct.
        target: The target to autoscale.
        min_allowed_cpu: The minimum allowed CPU in millicores. Defaults to 100.
        update_mode: The update mode to use. Defaults to "auto".

    Returns:
        A VerticalPodAutoscaler.
    """

    if update_mode == "off":
        update_mode = VerticalPodAutoscalerSpecUpdatePolicyUpdateMode.OFF
    elif update_mode == "initial":
        update_mode = VerticalPodAutoscalerSpecUpdatePolicyUpdateMode.INITIAL
    elif update_mode == "recreate":
        update_mode = VerticalPodAutoscalerSpecUpdatePolicyUpdateMode.RECREATE
    else:
        update_mode = VerticalPodAutoscalerSpecUpdatePolicyUpdateMode.AUTO

    return VerticalPodAutoscaler(
        scope,
        id,
        spec=VerticalPodAutoscalerSpec(
            target_ref=VerticalPodAutoscalerSpecTargetRef(
                api_version=target.api_version,
                kind=target.kind,
                name=target.name,
            ),
            update_policy=VerticalPodAutoscalerSpecUpdatePolicy(update_mode=update_mode),
            resource_policy=VerticalPodAutoscalerSpecResourcePolicy(
                container_policies=[
                    VerticalPodAutoscalerSpecResourcePolicyContainerPolicies(
                        container_name=name,
                        min_allowed={
                            "cpu": VerticalPodAutoscalerSpecResourcePolicyContainerPoliciesMinAllowed.from_number(
                                min_allowed_cpu
                            )
                        },
                    )
                    for name in target.containers
                ]
            ),
        ),
    )
