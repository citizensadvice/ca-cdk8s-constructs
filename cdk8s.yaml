# Defines the imports for CRDs
language: python
output: ca_cdk8s_constructs/imports
imports:
  - https://github.com/kubernetes/autoscaler/raw/refs/heads/master/vertical-pod-autoscaler/deploy/vpa-v1-crd-gen.yaml
  - https://raw.githubusercontent.com/external-secrets/external-secrets/refs/heads/main/deploy/crds/bundle.yaml