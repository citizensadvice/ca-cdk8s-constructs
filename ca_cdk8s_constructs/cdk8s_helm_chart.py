import builtins
from textwrap import dedent
import typing
from aws_cdk import Duration
from aws_cdk.aws_eks import HelmChart, ICluster
from aws_cdk.aws_s3_assets import Asset
import cdk8s
from constructs import Construct
import yaml
from tempfile import TemporaryDirectory
from pathlib import Path
from cdk8s_plus_32 import Job
from enum import Enum
from typing import Sequence


class HelmChartAsset(Construct):
    """
    Asset for a Helm chart that renders anything passed to `Values.resources` as a helm template
    """

    def __init__(
        self,
        scope: Construct,
        id: str,
        name: str,
        version: str,
        app_version: str,
    ) -> None:
        """
        Create an S3 asset for a cdk8s chart.
        """

        super().__init__(scope, id)

        # Create a temporary directory for the chart
        with TemporaryDirectory() as chart_folder:
            # Write the chart spec to the Chart.yaml file
            with open(Path(chart_folder, "Chart.yaml"), "w+") as chart_spec:
                chart_spec.write(
                    yaml.dump(
                        {
                            "apiVersion": "v2",
                            "name": name,
                            "description": f"Generated chart for {name}",
                            "type": "application",
                            "version": version,
                            "appVersion": app_version,
                        }
                    )
                )

            # Write the values.yaml file. This is a single key which will contain
            # the rendered cdk8s chart.
            with open(Path(chart_folder, "values.yaml"), "w+") as values:
                values.write("resources: {}")

            # Create the templates directory
            template_dir = Path(chart_folder, "templates")
            template_dir.mkdir(parents=True)

            # Write the resources.yaml file. This will contain the rendered cdk8s chart.
            with open(Path(template_dir, "resources.yaml"), "w+") as template:
                template.write(
                    dedent(
                        """\
                        {{- range $v := .Values.resources }}
                        ---
                        {{$v | toYaml }}
                        {{- end }}
                        """
                    )
                )

            # Create the chart asset
            self.asset = Asset(self, "Default", path=chart_folder)


class Cdk8sHelmChart(HelmChart):
    """
    Bundles a cdk8s chart as a Helm chart by rendering all chart resources into a single values item.

    Is used in the same way as the `aws-cdk.aws_eks.HelmChart` resource, but instead of `chart` or
    `chart_asset`, pass an instance of `cdk8s.Chart` to `cdk8s_chart`. This can then be deployed using
    the `aws-cdk.aws_eks.Cluster.add_helm_chart()` method and retains the usual Helm functionality.
    """

    def __init__(
        self,
        scope: Construct,
        id: str,
        *,
        chart_name: str,
        app_version: str,
        cluster: ICluster,
        cdk8s_chart: cdk8s.Chart,
        atomic: typing.Optional[builtins.bool] = None,
        create_namespace: typing.Optional[builtins.bool] = None,
        release: typing.Optional[builtins.str] = None,
        skip_crds: typing.Optional[builtins.bool] = None,
        timeout: typing.Optional[Duration] = None,
        version: typing.Optional[builtins.str] = None,
        wait: typing.Optional[builtins.bool] = None,
    ) -> None:
        chart_asset = HelmChartAsset(
            scope,
            "Package",
            name=chart_name,
            version=version or "1.0",
            app_version=app_version,
        ).asset

        super().__init__(
            scope,
            id,
            cluster=cluster,
            chart_asset=chart_asset,
            atomic=atomic,
            create_namespace=create_namespace,
            namespace=cdk8s_chart.namespace,
            release=release,
            skip_crds=skip_crds,
            timeout=timeout,
            values={"resources": cdk8s_chart.to_json()},
            wait=wait,
        )


class HelmHookType(Enum):
    """The type of hook to apply.

    pre-install: Executes after templates are rendered, but before any resources are created in Kubernetes
    post-install: Executes after all resources are loaded into Kubernetes
    pre-delete: Executes on a deletion request before any resources are deleted from Kubernetes
    post-delete: Executes on a deletion request after all of the release's resources have been deleted
    pre-upgrade: Executes on an upgrade request after templates are rendered, but before any resources are updated
    post-upgrade: Executes on an upgrade request after all resources have been upgraded
    pre-rollback: Executes on a rollback request after templates are rendered, but before any resources are rolled back
    post-rollback: Executes on a rollback request after all resources have been modified
    test: Executes when the Helm test subcommand is invoked
    """

    PRE_INSTALL = "pre-install"
    POST_INSTALL = "post-install"
    PRE_DELETE = "pre-delete"
    POST_DELETE = "post-delete"
    PRE_UPGRADE = "pre-upgrade"
    POST_UPGRADE = "post-upgrade"
    PRE_ROLLBACK = "pre-rollback"
    POST_ROLLBACK = "post-rollback"
    TEST = "test"


class HelmHookDeletePolicy(Enum):
    """The policy for deleting the hook resource after the hook has been executed.

    before-hook-creation: Delete the previous resource before a new hook is launched
    hook-succeeded: Delete the resource after the hook is successfully executed
    hook-failed: Delete the resource if the hook failed during execution
    """

    HOOK_SUCCEEDED = "hook-succeeded"
    BEFORE_HOOK_CREATION = "before-hook-creation"
    HOOK_FAILED = "hook-failed"


def apply_helm_hook(
    job: Job,
    hook_types: Sequence[HelmHookType],
    hook_delete_policy: Sequence[HelmHookDeletePolicy] = [
        HelmHookDeletePolicy.BEFORE_HOOK_CREATION
    ],
    weight: int = None,
) -> None:
    """Apply helm hook annotations to a job.

    This applies the correct annotations to the job to ensure that the job is executed as
    a helm hook if the resource is deployed as a helm chart.

    Args:
        job: The job to apply the helm hook to.
        hook_types: The types of hook to apply.
        hook_delete_policy: The policy for deleting the hook resource after the hook has been executed.
        weight: The weight of the hook.
    """
    job.metadata.add_annotation(
        "helm.sh/hook-delete-policy", ",".join([policy.value for policy in hook_delete_policy])
    )

    job.metadata.add_annotation("helm.sh/hook", ",".join([hook.value for hook in hook_types]))

    if weight:
        job.metadata.add_annotation("helm.sh/hook-weight", str(weight))
