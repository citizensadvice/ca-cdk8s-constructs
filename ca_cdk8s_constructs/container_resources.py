from cdk8s_plus_31 import ContainerResources, CpuResources, Cpu, MemoryResources
from cdk8s import Size
from numbers import Number


def ca_container_resources(cpu: Number, memory: Number):
    """Create a ContainerResources object.

    It is not recommended to set cpu limit so this value is not set by this construct.
    The memory limit will be set equal to the memory request.

    Args:
        cpu: The CPU requests in millicores (1000 = 1 CPU core).
        memory: The memory request in mebibytes (1024 = 1 GiB).
    """

    if cpu < 100:
        raise ValueError("CPU requests must be greater than or equal to 100 millicores")
    if memory < 256:
        raise ValueError(
            "Memory requests must be greater than or equal to 256 mebibytes"
        )

    return ContainerResources(
        cpu=CpuResources(
            request=Cpu.millis(amount=cpu),
        ),
        memory=MemoryResources(
            limit=Size.mebibytes(memory),
            request=Size.mebibytes(memory),
        ),
    )
