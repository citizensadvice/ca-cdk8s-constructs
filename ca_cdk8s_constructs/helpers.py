"""A collection of helper functions for cdk8s applications."""

from cdk8s import ApiObjectMetadata, ApiObjectMetadataDefinition
from constructs import Construct


def add_labels(construct: Construct, labels: dict[str, str]) -> None:
    """
    Recursively add the supplied labels to all resources in
    the construct and all its children that have a metadata
    property.

    args:
        construct: The construct to add labels to recursively.
        labels: The labels to add to the construct.
    """

    for resource in construct.node.children:
        if hasattr(resource, "metadata") and isinstance(
            resource.metadata, (ApiObjectMetadataDefinition, ApiObjectMetadata)
        ):
            for key, value in labels.items():
                resource.metadata.add_label(key, value)
        add_labels(resource, labels)
