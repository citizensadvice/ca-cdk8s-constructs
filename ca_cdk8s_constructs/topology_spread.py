from enum import Enum
from typing import Sequence

from cdk8s import JsonPatch
from cdk8s_plus_32 import Deployment, StatefulSet, Topology


class WhenUnsatisfiable(Enum):
    """Specifies how to deal with a pod if it doesn't satisfy the spread constraint.

    DO_NOT_SCHEDULE: Don't schedule the pod at all.
    SCHEDULE_ANYWAY: Schedule the pod anyway.
    """

    DO_NOT_SCHEDULE = "DoNotSchedule"
    SCHEDULE_ANYWAY = "ScheduleAnyway"


def set_topology_spread(
    target: Deployment | StatefulSet,
    topology_keys: Sequence[Topology] = (Topology.ZONE, Topology.HOSTNAME),
    when_unsatisfiable: WhenUnsatisfiable = WhenUnsatisfiable.SCHEDULE_ANYWAY,
    max_skew: int = 1,
    min_domains: int = 3,
):
    """Set topology spread constraints on a Deployment or StatefulSet.

    This function applies a topology spread constraint to a Deployment or
    StatefulSet to ensure pods are distributed across different Availability
    Zones and Nodes to improve high availability and fault tolerance.

    This function will replace any existing topology spread constraints on the
    Deployment or StatefulSet.

    It will set the maxSkew and minDomains to the provided values for all
    topology keys.

    Args:
        target: The Deployment or StatefulSet to apply the topology spread constraint to.
        topology_keys: The topology keys to spread the pods across.
        when_unsatisfiable: What to do if the pod doesn't satisfy the spread constraint.
        max_skew: The max difference between the number of pods in the most populated domain and the least populated domain.
        min_domains: The minimum number of domains to spread the pods across.
    """

    target._api_object.add_json_patch(
        JsonPatch.replace(
            path="/spec/template/spec/topologySpreadConstraints",
            value=[
                {
                    "maxSkew": max_skew,
                    "minDomains": min_domains,
                    "topologyKey": topology_key.key,
                    "whenUnsatisfiable": when_unsatisfiable.value,
                    "labelSelector": {
                        "matchLabels": target.match_labels,
                    },
                }
                for topology_key in topology_keys
            ],
        )
    )
