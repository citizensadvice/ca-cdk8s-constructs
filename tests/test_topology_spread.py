from cdk8s import Chart
from cdk8s_plus_32 import Deployment, Topology

from ca_cdk8s_constructs.topology_spread import WhenUnsatisfiable, set_topology_spread


def test_topology_spread_defaults(chart_fixture: Chart, snapshot):
    deployment = Deployment(chart_fixture, "Deployment")
    deployment.add_container(
        image="nginx:latest",
    )
    set_topology_spread(
        target=deployment,
    )
    constraints = chart_fixture.to_json()[0]["spec"]["template"]["spec"][
        "topologySpreadConstraints"
    ]
    assert constraints == snapshot


def test_topology_spread_max_skew(chart_fixture: Chart, snapshot):
    deployment = Deployment(chart_fixture, "Deployment")
    deployment.add_container(
        image="nginx:latest",
    )
    set_topology_spread(
        target=deployment,
        max_skew=2,
    )
    constraints = chart_fixture.to_json()[0]["spec"]["template"]["spec"][
        "topologySpreadConstraints"
    ]
    assert constraints == snapshot


def test_topology_spread_min_domains(chart_fixture: Chart, snapshot):
    deployment = Deployment(chart_fixture, "Deployment")
    deployment.add_container(
        image="nginx:latest",
    )
    set_topology_spread(
        target=deployment,
        min_domains=2,
    )
    constraints = chart_fixture.to_json()[0]["spec"]["template"]["spec"][
        "topologySpreadConstraints"
    ]
    assert constraints == snapshot


def test_topology_spread_when_unsatisfiable(chart_fixture: Chart, snapshot):
    deployment = Deployment(chart_fixture, "Deployment")
    deployment.add_container(
        image="nginx:latest",
    )
    set_topology_spread(
        target=deployment,
        when_unsatisfiable=WhenUnsatisfiable.DO_NOT_SCHEDULE,
    )
    constraints = chart_fixture.to_json()[0]["spec"]["template"]["spec"][
        "topologySpreadConstraints"
    ]
    assert constraints == snapshot


def test_topology_spread_topology_keys(chart_fixture: Chart, snapshot):
    deployment = Deployment(chart_fixture, "Deployment")
    deployment.add_container(
        image="nginx:latest",
    )
    set_topology_spread(
        target=deployment,
        topology_keys=(Topology.ZONE,),
    )
    constraints = chart_fixture.to_json()[0]["spec"]["template"]["spec"][
        "topologySpreadConstraints"
    ]
    assert constraints == snapshot
