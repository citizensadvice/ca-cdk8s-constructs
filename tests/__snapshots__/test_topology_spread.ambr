# serializer version: 1
# name: test_topology_spread_defaults
  list([
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 3,
      'topologyKey': 'topology.kubernetes.io/zone',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 3,
      'topologyKey': 'kubernetes.io/hostname',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
  ])
# ---
# name: test_topology_spread_max_skew
  list([
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 2,
      'minDomains': 3,
      'topologyKey': 'topology.kubernetes.io/zone',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 2,
      'minDomains': 3,
      'topologyKey': 'kubernetes.io/hostname',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
  ])
# ---
# name: test_topology_spread_min_domains
  list([
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 2,
      'topologyKey': 'topology.kubernetes.io/zone',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 2,
      'topologyKey': 'kubernetes.io/hostname',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
  ])
# ---
# name: test_topology_spread_topology_keys
  list([
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 3,
      'topologyKey': 'topology.kubernetes.io/zone',
      'whenUnsatisfiable': 'ScheduleAnyway',
    }),
  ])
# ---
# name: test_topology_spread_when_unsatisfiable
  list([
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 3,
      'topologyKey': 'topology.kubernetes.io/zone',
      'whenUnsatisfiable': 'DoNotSchedule',
    }),
    dict({
      'labelSelector': dict({
        'matchLabels': dict({
          'cdk8s.io/metadata.addr': 'test-Deployment-c83f5e59',
        }),
      }),
      'maxSkew': 1,
      'minDomains': 3,
      'topologyKey': 'kubernetes.io/hostname',
      'whenUnsatisfiable': 'DoNotSchedule',
    }),
  ])
# ---
